en/dev_log/ulog_file_format.md

```diff
@@ -41,233 +51,332 @@ The file consists of three sections:
 ----------------------
 ```
 
+A description of each section is provided below.
+
+
 ### Header Section
 
 The header is a fixed-size section and has the following format (16 bytes):
+
 ```
 ----------------------------------------------------------------------
 | 0x55 0x4c 0x6f 0x67 0x01 0x12 0x35 | 0x01         | uint64_t       |
 | File magic (7B)                    | Version (1B) | Timestamp (8B) |
 ----------------------------------------------------------------------
 ```
 
-Version is the file format version, currently 1.
-Timestamp is a `uint64_t` integer, denotes the start of the logging in microseconds.
+- **File Magic (7 Bytes):** File type indicator that reads "ULogXYZ where XYZ is the magic bytes sequence `0x01 0x12 0x35`"
+- **Version (1 Byte):** File format version (currently 1)
+- **Timestamp (8 Bytes):** `uint64_t` integer that denotes when the logging started in microseconds.
+
+### Definition & Data Section Message Header
+
+The _Definitions and Data_ sections contain a number of **messages**. Each message is preceded by this header:
+
+```c
+struct message_header_s {
+  uint16_t msg_size;
+  uint8_t msg_type;
+};
+```
+
+- `msg_size` is the size of the message in bytes without the header.
+- `msg_type` defines the content, and is a single alphabet character.
+
+In practice, we include the `msg_size` and `msg_type` in each message definitions as the first 2 members of different message definitions.
+
+:::note
+Messages below are prefixed with the single alphabet that corresponds to it's `msg_type`.
+:::
 
 
 ### Definitions Section
 
-Variable length section, contains version information, format definitions, and (initial) parameter values.
+The definitions section contains basic information such as software version, uORB Topic format, initial parameter values, and so on.
+
+The message types in this section are:
+1. [Flag Bits](#b-flag-bits-message)
+2. [Format Definition](#f-format-message)
+3. [Information](#i-information-message)
+4. [Multi Information](#m-multi-information-message)
+5. [Parameter](#p-parameter-message)
+6. [Default Parameter](#q-default-parameter-message)
+
+
+#### 'B': Flag Bits Message (`msg_type = 'B'`)
+
+:::note
+This message must be the **first message** right after the header section, so that it has a fixed constant offset from the start of the file!
+:::
 
-The Definitions and Data sections consist of a stream of messages.
-Each starts with this header:
 ```c
-struct message_header_s {
-	uint16_t msg_size;
-	uint8_t msg_type
+struct ulog_message_flag_bits_s {
+  uint16_t msg_size;
+  uint8_t msg_type = 'B';
+  uint8_t compat_flags[8];
+  uint8_t incompat_flags[8];
+  uint64_t appended_offsets[3]; ///< file offset(s) for appended data if appending bit is set
 };
 ```
 
-`msg_size` is the size of the message in bytes without the header (`hdr_size`= 3 bytes).
-`msg_type` defines the content and is one of the following:
+- `compat_flags`: compatible flag bits
```

Comment:
> How do we explain the compat_flags and incompat_flags more clearly? -> https://github.com/PX4/PX4-user_guide/pull/1887#issuecomment-1137202556

I don't find this confusing, perhaps because the concept is also used in mavlink.

Essentially there are two types of changes you might make to protocol or log or whatever - compatible or incompatible. 

An incompatible change is one that makes the content usuable - you simply have to throw away everything in the scope of the change.
- In mavlink this might be something "like" compression of a message - if a parser were to see a compression incompatibility bit set and it didn't know what that meant it would have to drop the message
- in the log this means you have to discard the whole log if you don't implement the feature indicated by an incompatible bit. This might mean the indicated extension is essential to using the log. I "guess" it could also be like the mavlink case, which means it makes the log unparsable. 

A compatible change is one that a parser can just ignore - it should continue to parse as normal - the output might not include some information that is in the log, but that isn't essential, and the parser can just go it's merry way.
For mavlink this is something like a bit that indicates a particular message is high priority. If a parser doesn't implement high priority parsing and the bit is appended to the end so that parsers that don't understand it won't even see it - then it is compatible. For the log this is just "default parameters message" - you can use a log even if you just ignore this message.

Anyway, leave to you to decide if  more needs to be said.



What this says is that if any of the `compat_flags` flags are set a parser that d

Reply:
Thank you for the explanation! It was applied in the latest commit